tf_helpers/lax.py

```diff
@@ -0,0 +1,505 @@
+    else:
+      raise ValueError('No 4+ dimensional dimension_number defaults.')
+  dn = conv_dimension_numbers(lhs.shape, rhs.shape, dimension_numbers)
+  k_shape = onp.take(rhs.shape, dn.rhs_spec)
+  k_sdims = k_shape[2:]
+  # Calculate correct output shape given padding and strides.
+  pads: Union[str, Sequence[Tuple[int, int]]]
+  if padding in {'SAME', 'VALID'}:
+    if rhs_dilation is None:
+      rhs_dilation = (1,) * (rhs.ndim - 2)
+    effective_k_size = map(lambda k, r: (k-1) * r + 1, k_sdims, rhs_dilation)
+    pads = [_conv_transpose_padding(k, s, padding)
+            for k,s in zip(effective_k_size, strides)]
+  else:
+    pads = padding
+  if transpose_kernel:
+    # flip spatial dims and swap input / output channel axes
+    rhs = _flip_axes(rhs, onp.array(dn.rhs_spec)[2:])
+    rhs = onp.swapaxes(rhs, dn.rhs_spec[0], dn.rhs_spec[1])
+  # return conv_general_dilated(lhs, rhs, one, pads, strides, rhs_dilation, dn,
+  #                             precision=precision)
+  return conv_general_dilated(lhs, rhs, one, pads, strides, rhs_dilation, dn)
+
+
+def dot_general(lhs, rhs, dimension_numbers, precision=None):
+  """ The general dot operation for TensorFlow.
+
+  An equivalent general dot operation as that in JAX -
+     <https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.dot_general.html>
+  Although there is an implementation in TF XLA, avoid directly using XLA when
+  possible.
+
+  e.g., non-batched: ij,jk->ik
+        batched: ijk,ikl->ijl
+
+  Args:
+    lhs: an array (the left-hand side matrix/vector to be multiplied)
+    rhs: an array (the right-hand side matrix/vector to be multiplied)
+    dimension_numbers: (Tuple[Tuple[Sequence[int], Sequence[int]],
+      Tuple[Sequence[int], Sequence[int]]]) – a tuple of tuples of the form
+      ((lhs_contracting_dims, rhs_contracting_dims), (lhs_batch_dims, rhs_batch_dims))
+
+  Returns:
+    An array that contains the result.
+  """
+  char_list = list(string.ascii_lowercase)
+  char_list = char_list[8:] + char_list[:8]
+  lhs_rank, rhs_rank = len(lhs.shape), len(rhs.shape)
+  lhs_rep = char_list[:lhs_rank]
+  rhs_rep = char_list[lhs_rank:lhs_rank+rhs_rank]
+  contraction, batch = dimension_numbers
+  lhs_contraction, rhs_contraction = contraction
+  if len(lhs_contraction) != len(rhs_contraction):
+    raise ValueError("The input matrices are required to have the same number "
+                     "of contraction dimensions, but got: lhs {}, rhs: {}".format(
+                     len(lhs_contraction), len(rhs_contraction)))
+  lhs_batch, rhs_batch = batch
+  if len(lhs_batch) != len(rhs_batch):
+    raise ValueError("The input matrices are required to have the same number "
+                     "of batch dimensions, but got: lhs {}, rhs: {}".format(
+                     len(lhs_batch), len(rhs_batch)))
+
+  if len(lhs_batch) == 0 and len(rhs_batch) == 0:
+    return _non_batched_matmul(lhs, rhs, lhs_contraction, rhs_contraction)
```

Comment:
I still feel like this call doesn't justify a whole separate function, as opposed to e.g. the `linalg.matmul` call below. Unless it's gonna be used a lot elsewhere. Lmk if you think otherwise!

Reply:
Hi Roman, I did not quite get what you mean here. Would you mind elaborating on it?

Actually this functionality has already been merged into Google Trax repo - https://github.com/google/trax/pull/956.

Let me know!

Reply:
I meant that function `_non_batched_matmul` doesn't seem to do much but just calls `tf.tensordot`, so I'm not sure if this needs to be a separate functions instead of just calling `tf.tensordot`, like you do below when you call `tf.linalg.matmul`. But it's a minor point, OK either way!

Reply:
Oh yes, you are right! At the beginning this helper did not look like this, and this is the one after many iterations of editing.

I will remove this helper as you suggested.